python/ray/serve/pipeline/api.py

```diff
@@ -57,5 +57,12 @@ def build(ray_dag_root_node: DAGNode):
         >>> deployments = serve.pipeline.build(m1) # or just a regular node.
     """
     serve_root_dag = ray_dag_root_node.apply_recursive(transform_ray_dag_to_serve_dag)
-    deployments = extract_deployments_from_serve_dag(serve_root_dag)
-    return deployments
+    body_deployments, root_deployment = extract_deployments_from_serve_dag(
+        serve_root_dag
+    )
+    # Found user facing DeploymentNode / DeploymentFunctionNode as ingress
+    if len(root_deployment) == 1:
```

Comment:
at most one root can be returned and we append to final return list so existing ray dag tests are compatible.

We only do the prefix and # of root deployment assertion at `serve.run` since that's the only place we add and call http.

Reply:
hmm I'm confused about this, why don't we do it in the DAG building step to keep it handled in one layer? we are going to need the same logic for `serve build` as well

Reply:
hmm by saying "DAG building step" you mean the `pipeline_bulid` here just turns everything needed instead of requiring additional processing ? To be clear this function is already called for current `serve.run`, but there're tests who are only testing the python DAG executability without knowledge of, or test with http. 

I think you're suggesting to move the `serve/api.py` assertions back here ?

Reply:
i think i see what you're saying after reading the build PR. it's simpler if `pipeline_build` just returns a list of deployments. For setting route prefix we should only do it at upper layer tho since `pipeline_build`'s primary responsibility is to transform and produce a runnable serve dag and return all deployment uses, without knowledge of http.